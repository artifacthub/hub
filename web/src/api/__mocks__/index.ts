export const API = {
  getPackage: jest.fn(),
  toggleStar: jest.fn(),
  getStars: jest.fn(),
  searchPackages: jest.fn(),
  getStats: jest.fn(),
  getRandomPackages: jest.fn(),
  register: jest.fn(),
  verifyEmail: jest.fn(),
  login: jest.fn(),
  logout: jest.fn(),
  getUserProfile: jest.fn(),
  getAllRepositories: jest.fn(),
  getRepositories: jest.fn(),
  addRepository: jest.fn(),
  deleteRepository: jest.fn(),
  updateRepository: jest.fn(),
  transferRepository: jest.fn(),
  claimRepositoryOwnership: jest.fn(),
  checkAvailability: jest.fn(),
  getUserOrganizations: jest.fn(),
  getOrganization: jest.fn(),
  addOrganization: jest.fn(),
  updateOrganization: jest.fn(),
  deleteOrganization: jest.fn(),
  getOrganizationMembers: jest.fn(),
  addOrganizationMember: jest.fn(),
  deleteOrganizationMember: jest.fn(),
  confirmOrganizationMembership: jest.fn(),
  getStarredByUser: jest.fn(),
  updateUserProfile: jest.fn(),
  updatePassword: jest.fn(),
  saveImage: jest.fn(),
  getPackageSubscriptions: jest.fn(),
  addSubscription: jest.fn(),
  deleteSubscription: jest.fn(),
  getUserSubscriptions: jest.fn(),
  getWebhooks: jest.fn(),
  getWebhook: jest.fn(),
  addWebhook: jest.fn(),
  deleteWebhook: jest.fn(),
  updateWebhook: jest.fn(),
  triggerWebhookTest: jest.fn(),
  getAPIKeys: jest.fn(),
  getAPIKey: jest.fn(),
  addAPIKey: jest.fn(),
  updateAPIKey: jest.fn(),
  deleteAPIKey: jest.fn(),
  getOptOutList: jest.fn(),
  addOptOut: jest.fn(),
  deleteOptOut: jest.fn(),
  getAuthorizationPolicy: jest.fn(),
  updateAuthorizationPolicy: jest.fn(),
  getUserAllowedActions: jest.fn(),
  getSnapshotSecurityReport: jest.fn(),
  getValuesSchema: jest.fn(),
  getChangelog: jest.fn(),
  triggerTestInRegoPlayground: jest.fn(),
};
