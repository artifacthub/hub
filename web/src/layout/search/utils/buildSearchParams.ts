import { Filters } from '../../../types';
import isNull from 'lodash/isNull';

const SPECIAL_KEYS = ['text', 'page'];

interface F {
  [key: string]: string[];
}

export default (query: string): Filters => {
  const p = new URLSearchParams(query);
  let f: F = {};

  p.forEach((value, key) => {
    if (!SPECIAL_KEYS.includes(key)) {
      const values = f[key] || [];
      values.push(value);
      f[key] = values;
    }
  });

  return {
    text: p.has('text') ? p.get('text')! : undefined,
    f: { ...f },
    pageNumber: p.has('page') && !isNull(p.get('page')) ? parseInt(p.get('page')!) : 1,
  };
}
