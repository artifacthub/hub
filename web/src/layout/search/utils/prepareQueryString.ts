import isUndefined from 'lodash/isUndefined';
import { Filters } from '../../../types';

export default (filters: Filters): string => {
  const p = new URLSearchParams();
  p.set('page', filters.pageNumber.toString());
  if (!isUndefined(filters.text)) {
    p.set('text', filters.text);
  }
  if (!isUndefined(filters.f)) {
    Object.keys(filters.f).forEach((filterId: string) => {
      return filters.f[filterId].forEach((id: string) => {
        p.append(filterId, id);
      });
    });
  }
  const result = p.toString();
  return result === '' ? '' : `?${result}`;
}
