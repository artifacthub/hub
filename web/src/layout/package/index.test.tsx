import { fireEvent, render, waitFor } from '@testing-library/react';
import React from 'react';
import { BrowserRouter as Router } from 'react-router-dom';
import { mocked } from 'ts-jest/utils';

import { API } from '../../api';
import { ErrorKind, Package } from '../../types';
import prepareQuerystring from '../../utils/prepareQueryString';
import PackageView from './index';
jest.mock('../../api');
jest.mock('../../utils/updateMetaIndex');

const getMockPackage = (fixtureId: string): Package => {
  return require(`./__fixtures__/index/${fixtureId}.json`) as Package;
};

const mockHistoryPush = jest.fn();
const mockHistoryReplace = jest.fn();

jest.mock('react-router-dom', () => ({
  ...(jest.requireActual('react-router-dom') as {}),
  useHistory: () => ({
    push: mockHistoryPush,
    replace: mockHistoryReplace,
    action: 'POP',
  }),
}));

const defaultProps = {
  repositoryKind: 'helm',
  repositoryName: 'repoName',
  packageName: 'packageName',
  searchUrlReferer: undefined,
};

describe('Package index', () => {
  afterEach(() => {
    jest.resetAllMocks();
  });

  it('creates snapshot', async () => {
    const mockPackage = getMockPackage('1');
    mocked(API).getPackage.mockResolvedValue(mockPackage);

    const result = render(
      <Router>
        <PackageView {...defaultProps} />
      </Router>
    );

    await waitFor(() => {
      expect(API.getPackage).toHaveBeenCalledTimes(1);
      expect(result.asFragment()).toMatchSnapshot();
    });
  });

  describe('Render', () => {
    it('renders component', async () => {
      const mockPackage = getMockPackage('2');
      mocked(API).getPackage.mockResolvedValue(mockPackage);

      render(
        <Router>
          <PackageView {...defaultProps} />
        </Router>
      );

      await waitFor(() => {
        expect(API.getPackage).toHaveBeenCalledTimes(1);
      });
    });

    it('displays loading spinner', async () => {
      const mockPackage = getMockPackage('3');
      mocked(API).getPackage.mockResolvedValue(mockPackage);

      const props = {
        ...defaultProps,
      };

      const { getByRole } = render(
        <Router>
          <PackageView {...props} />
        </Router>
      );

      await waitFor(() => {
        expect(getByRole('status')).toBeTruthy();
      });
    });
  });

  describe('when getPackage fails', () => {
    it('generic error', async () => {
      mocked(API).getPackage.mockRejectedValue({
        kind: ErrorKind.Other,
      });

      const props = {
        ...defaultProps,
      };

      const { getByTestId } = render(
        <Router>
          <PackageView {...props} />
        </Router>
      );

      await waitFor(() => {
        expect(API.getPackage).toHaveBeenCalledTimes(1);
      });

      const noData = getByTestId('noData');
      expect(noData).toBeInTheDocument();
      expect(noData).toHaveTextContent(/An error occurred getting this package, please try again later./i);
    });

    it('not found package', async () => {
      mocked(API).getPackage.mockRejectedValue({
        kind: ErrorKind.NotFound,
      });

      const props = {
        ...defaultProps,
      };

      const { getByTestId, getByText } = render(
        <Router>
          <PackageView {...props} />
        </Router>
      );

      await waitFor(() => {
        expect(API.getPackage).toHaveBeenCalledTimes(1);
      });

      const noData = getByTestId('noData');
      expect(noData).toBeInTheDocument();
      expect(noData).toHaveTextContent('Sorry, the package you requested was not found.');
      expect(
        getByText(
          'The package you are looking for may have been deleted by the provider, or it may now belong to a different repository. Please try searching for it, as it may help locating the package in a different repository or discovering other alternatives.'
        )
      ).toBeInTheDocument();
    });
  });

  describe('Go back button', () => {
    it('proper behaviour', async () => {
      const searchUrlReferer = {
        tsQueryWeb: 'test',
        filters: {},
        pageNumber: 1,
        deprecated: false,
      };
      const mockPackage = getMockPackage('4');
      mocked(API).getPackage.mockResolvedValue(mockPackage);

      const { getByTestId } = render(
        <Router>
          <PackageView {...defaultProps} searchUrlReferer={searchUrlReferer} />
        </Router>
      );

      const goBack = await waitFor(() => getByTestId('goBack'));
      expect(goBack).toBeInTheDocument();
      fireEvent.click(goBack);
      expect(mockHistoryPush).toHaveBeenCalledTimes(1);
      expect(mockHistoryPush).toHaveBeenCalledWith({
        pathname: '/packages/search',
        search: prepareQuerystring(searchUrlReferer),
        state: { 'from-detail': true },
      });
    });
  });

  describe('Repository button', () => {
    it('renders repository link', async () => {
      const mockPackage = getMockPackage('5');
      mocked(API).getPackage.mockResolvedValue(mockPackage);

      const { getByTestId } = render(
        <Router>
          <PackageView {...defaultProps} />
        </Router>
      );

      const link = await waitFor(() => getByTestId('repoLink'));
      expect(link).toBeInTheDocument();
      fireEvent.click(link);
      expect(mockHistoryPush).toHaveBeenCalledTimes(1);
      expect(mockHistoryPush).toHaveBeenCalledWith({
        pathname: '/packages/search',
        search: prepareQuerystring({
          pageNumber: 1,
          filters: {
            repo: [mockPackage.repository.name],
          },
          deprecated: mockPackage.deprecated,
        }),
      });
    });
  });

  describe('Modal', () => {
    it('renders properly', async () => {
      const mockPackage = getMockPackage('6');
      mocked(API).getPackage.mockResolvedValue(mockPackage);

      const { getByRole } = render(
        <Router>
          <PackageView {...defaultProps} />
        </Router>
      );

      await waitFor(() => {
        expect(getByRole('dialog')).toBeInTheDocument();
      });
    });
  });

  describe('Readme', () => {
    it('does not render it when readme is null and displays no data message', async () => {
      const mockPackage = getMockPackage('7');
      mocked(API).getPackage.mockResolvedValue(mockPackage);

      const { getByTestId, queryByTestId } = render(
        <Router>
          <PackageView {...defaultProps} />
        </Router>
      );

      const noData = await waitFor(() => getByTestId('noData'));
      expect(noData).toBeInTheDocument();
      expect(noData).toHaveTextContent('No README file available for this package');
      expect(queryByTestId('readme')).toBeNull();
    });

    it('renders it correctly', async () => {
      const mockPackage = getMockPackage('8');
      mocked(API).getPackage.mockResolvedValue(mockPackage);

      const { queryByTestId } = render(
        <Router>
          <PackageView {...defaultProps} />
        </Router>
      );

      await waitFor(() => {
        expect(API.getPackage).toHaveBeenCalledTimes(1);
      });

      const readme = await waitFor(() => queryByTestId('readme'));
      expect(readme).toBeInTheDocument();
      expect(readme).toHaveTextContent(mockPackage.readme!);
    });
  });

  describe('Labels', () => {
    it('renders Verified Publisher label', async () => {
      const mockPackage = getMockPackage('9');
      mocked(API).getPackage.mockResolvedValue(mockPackage);

      const { getAllByText } = render(
        <Router>
          <PackageView {...defaultProps} />
        </Router>
      );

      await waitFor(() => {
        expect(API.getPackage).toHaveBeenCalledTimes(1);
      });

      expect(getAllByText('Verified Publisher')).toHaveLength(2);
    });
  });

  describe('Helm package', () => {
    it('renders CRDs when are defined', async () => {
      const mockPackage = getMockPackage('10');
      mocked(API).getPackage.mockResolvedValue(mockPackage);

      const { getByText, getByTestId } = render(
        <Router>
          <PackageView {...defaultProps} />
        </Router>
      );

      await waitFor(() => {
        expect(API.getPackage).toHaveBeenCalledTimes(1);
      });

      expect(getByText('Custom Resource Definitions')).toBeInTheDocument();
      expect(getByTestId('resourceDefinition')).toBeInTheDocument();
    });
  });

  describe('OLM package', () => {
    it('renders CRDs from crds prop when is defined', async () => {
      const mockPackage = getMockPackage('11');
      mocked(API).getPackage.mockResolvedValue(mockPackage);

      const { getByText, getAllByTestId } = render(
        <Router>
          <PackageView {...defaultProps} />
        </Router>
      );

      await waitFor(() => {
        expect(API.getPackage).toHaveBeenCalledTimes(1);
      });

      expect(getByText('Custom Resource Definitions')).toBeInTheDocument();
      expect(getAllByTestId('resourceDefinition')).toHaveLength(1);
    });
  });

  describe('Falco rules', () => {
    it('renders rules properly', async () => {
      const mockPackage = getMockPackage('12');
      mocked(API).getPackage.mockResolvedValue(mockPackage);

      const { getAllByTestId, getByTestId, getByText } = render(
        <Router>
          <PackageView {...defaultProps} />
        </Router>
      );

      await waitFor(() => {
        expect(API.getPackage).toHaveBeenCalledTimes(1);
      });

      expect(getByTestId('mainPackage')).toBeInTheDocument();
      expect(getByText('Rules')).toBeInTheDocument();
      expect(getAllByTestId('ctcBtn')).toHaveLength(2);
      expect(getByTestId('downloadBtn')).toBeInTheDocument();
    });
  });

  describe('Tekton task', () => {
    it('renders task properly', async () => {
      const mockPackage = getMockPackage('13');
      mocked(API).getPackage.mockResolvedValue(mockPackage);

      const { getByTestId, getByText } = render(
        <Router>
          <PackageView {...defaultProps} />
        </Router>
      );

      await waitFor(() => {
        expect(API.getPackage).toHaveBeenCalledTimes(1);
      });

      expect(getByText('Manifest YAML')).toBeInTheDocument();
      expect(getByTestId('tektonManifestBtn')).toBeInTheDocument();
    });

    describe('Krew kubectl plugin', () => {
      it('renders plugin properly', async () => {
        const mockPackage = getMockPackage('14');
        mocked(API).getPackage.mockResolvedValue(mockPackage);

        const { getByText, getAllByTestId, getByTestId } = render(
          <Router>
            <PackageView {...defaultProps} />
          </Router>
        );

        await waitFor(() => {
          expect(API.getPackage).toHaveBeenCalledTimes(1);
        });

        expect(getByText('Manifest')).toBeInTheDocument();
        expect(getAllByTestId('ctcBtn')).toHaveLength(2);
        expect(getByTestId('downloadBtn')).toBeInTheDocument();
      });
    });
  });
});
