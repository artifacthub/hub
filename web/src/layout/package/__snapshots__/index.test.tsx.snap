// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Package index creates snapshot 1`] = `
<DocumentFragment>
  <div
    class="position-relative flex-grow-1"
    data-testid="mainPackage"
  />
</DocumentFragment>
`;

exports[`Package index creates snapshot 2`] = `
<DocumentFragment>
  <div
    class="position-relative flex-grow-1"
    data-testid="mainPackage"
  >
    <div
      class="jumbotron package-detail-jumbotron jumbotron"
    >
      <div
        class="container position-relative"
      >
        <div
          class="d-flex align-items-start w-100 mb-3"
        >
          <div
            class="d-flex align-items-center flex-grow-1 mw-100"
          >
            <div
              class="d-flex align-items-center justify-content-center p-1 overflow-hidden imageWrapper imageWrapper"
            >
              <img
                alt="Pretty name"
                class="image"
                src="/image/imageId"
                srcset="/image/imageId@1x 1x, /image/imageId@2x 2x, /image/imageId@3x 3x, /image/imageId@4x 4x"
              />
            </div>
            <div
              class="ml-3 flex-grow-1 wrapperWithContentEllipsis"
            >
              <div
                class="d-flex flex-row align-items-center titleWrapper"
              >
                <div
                  class="h3 mb-0 text-nowrap text-truncate title"
                >
                  Pretty name
                </div>
                <div
                  class="d-none d-md-flex ml-3"
                />
              </div>
              <div
                class="d-flex d-md-none text-truncate mt-2 w-100 mobileSubtitle"
              >
                <small
                  class="text-muted text-uppercase"
                >
                  Repo: 
                </small>
                <div
                  class="mx-1 d-inline mobileIcon"
                >
                  <img
                    alt="Icon"
                    class="repoIcon repoIcon"
                    src="/static/media/helm-chart.svg"
                  />
                </div>
                <span
                  class="text-dark d-inline-block text-truncate mw-100 mobileVersion"
                >
                  incubator
                </span>
              </div>
              <div
                class="d-none d-md-flex flex-row align-items-baseline mt-2 subtitle"
              >
                <div
                  class="mr-2 text-truncate d-flex flex-row align-items-baseline mw50"
                >
                  <div
                    class="position-absolute"
                  >
                    <div
                      class="dropdown-menu dropdown-menu-left dropdown"
                      data-testid="orgInfoDropdown"
                    />
                  </div>
                  <div
                    class="d-flex flex-row aling-items-start text-truncate"
                  >
                    <div
                      class="d-flex flex-row align-items-baseline mr-1 text-muted text-uppercase"
                    >
                      <small>
                        Org:
                      </small>
                    </div>
                    <button
                      class="p-0 border-0 text-dark text-truncate flex-grow-1 link undefined"
                      data-testid="orgLink"
                    >
                      <div
                        class="text-truncate"
                      >
                        Helm
                      </div>
                    </button>
                  </div>
                </div>
                <div
                  class="text-truncate d-flex flex-row align-items-baseline mw50"
                >
                  <div
                    class="position-absolute"
                  >
                    <div
                      class="dropdown-menu dropdown-menu-left dropdown"
                      data-testid="repoInfoDropdown"
                    >
                      <div
                        class="content"
                      >
                        <div
                          class="d-flex flex-column"
                        >
                          <div
                            class="d-flex flex-row align-items-center"
                          >
                            <small
                              class="text-muted text-uppercase mr-1"
                            >
                              Repo: 
                            </small>
                            <img
                              alt="Icon"
                              class="mr-1 repoIconMini repoIcon"
                              src="/static/media/helm-chart.svg"
                            />
                            <div
                              class="text-reset labelContent"
                            >
                              incubator
                            </div>
                          </div>
                          <div
                            class="mt-2 d-flex flex-row align-items-baseline"
                          >
                            <small
                              class="text-muted text-uppercase mr-1"
                            >
                              Url: 
                            </small>
                            <div
                              class="text-reset urlContent labelContent"
                            >
                              <div
                                class="position-relative d-inline"
                              >
                                <button
                                  class="btn btn-sm miniBtn"
                                  data-testid="ctcBtn"
                                  type="button"
                                >
                                  <div
                                    class="d-flex flex-row align-items-center"
                                  >
                                    <svg
                                      fill="none"
                                      height="1em"
                                      stroke="currentColor"
                                      stroke-linecap="round"
                                      stroke-linejoin="round"
                                      stroke-width="2"
                                      viewBox="0 0 24 24"
                                      width="1em"
                                      xmlns="http://www.w3.org/2000/svg"
                                    >
                                      <rect
                                        height="13"
                                        rx="2"
                                        ry="2"
                                        width="13"
                                        x="9"
                                        y="9"
                                      />
                                      <path
                                        d="M5 15H4a2 2 0 0 1-2-2V4a2 2 0 0 1 2-2h9a2 2 0 0 1 2 2v1"
                                      />
                                    </svg>
                                  </div>
                                </button>
                              </div>
                            </div>
                          </div>
                        </div>
                      </div>
                    </div>
                  </div>
                  <div
                    class="d-flex flex-row aling-items-baseline text-truncate"
                  >
                    <div
                      class="d-flex flex-row align-items-baseline mr-1 text-muted text-uppercase"
                    >
                      <small>
                        Repo:
                      </small>
                      <img
                        alt="Icon"
                        class="ml-1 repoIcon repoIcon"
                        src="/static/media/helm-chart.svg"
                      />
                    </div>
                    <button
                      class="d-flex flex-row p-0 border-0 text-dark text-truncate link"
                      data-testid="repoLink"
                    >
                      <div
                        class="text-truncate"
                      >
                        incubator
                      </div>
                    </button>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
        <p
          class="mb-0 description"
        >
          desc
        </p>
        <div
          class="d-flex flex-wrap d-md-none"
        />
        <div
          class="position-absolute d-flex flex-row align-items-center optsWrapper"
        >
          <span
            class="d-block d-md-none text-muted text-nowrap date"
          >
            Updated Invalid date
          </span>
        </div>
        <div
          class="d-flex align-items-center justify-content-between flex-wrap d-md-none"
        >
          <div
            class="d-flex w-100 mt-3"
          >
            <div
              class="pr-1 w-50"
            >
              <div
                class="wrapper"
              >
                <div
                  class="position-relative buttonWrapper"
                >
                  <button
                    class="font-weight-bold text-uppercase position-relative btn btn-block btn btn-outline-secondary btn-sm"
                    data-testid="openModalBtn"
                    type="button"
                  >
                    <div
                      class="d-flex align-items-center justify-content-center"
                    >
                      <svg
                        class="mr-2"
                        fill="none"
                        height="1em"
                        stroke="currentColor"
                        stroke-linecap="round"
                        stroke-linejoin="round"
                        stroke-width="2"
                        viewBox="0 0 24 24"
                        width="1em"
                        xmlns="http://www.w3.org/2000/svg"
                      >
                        <line
                          x1="12"
                          x2="12"
                          y1="5"
                          y2="19"
                        />
                        <line
                          x1="5"
                          x2="19"
                          y1="12"
                          y2="12"
                        />
                      </svg>
                      <span>
                        Info
                      </span>
                    </div>
                  </button>
                </div>
                <div
                  class="modal modal"
                  role="dialog"
                >
                  <div
                    class="modal-dialog modal-lg modal-dialog-centered modal-dialog-scrollable"
                  >
                    <div
                      class="modal-content content undefined"
                    >
                      <div
                        class="modal-header d-flex flex-row align-items-center header undefined"
                      >
                        <div
                          class="d-flex align-items-center"
                        >
                          <div
                            class="d-flex align-items-center justify-content-center p-1 overflow-hidden imageWrapper imageWrapper"
                          >
                            <img
                              alt="Pretty name"
                              class="image"
                              src="/image/imageId"
                              srcset="/image/imageId@1x 1x, /image/imageId@2x 2x, /image/imageId@3x 3x, /image/imageId@4x 4x"
                            />
                          </div>
                          <div
                            class="ml-3"
                          >
                            <div
                              class="h5 mb-0"
                            >
                              Pretty name
                            </div>
                          </div>
                        </div>
                        <button
                          class="close"
                          data-testid="closeModalBtn"
                          type="button"
                        >
                          <span
                            aria-hidden="true"
                          >
                            ×
                          </span>
                        </button>
                      </div>
                      <div
                        class="modal-body p-4 h-100 d-flex flex-column"
                      >
                        <div
                          class="alertWrapper"
                          data-testid="alertWrapper"
                        />
                      </div>
                      <div
                        class="modal-footer p-3"
                      >
                        <button
                          class="btn btn-sm btn-secondary text-uppercase"
                          data-testid="closeModalFooterBtn"
                          type="button"
                        >
                          <div
                            class="d-flex flex-row align-items-center"
                          >
                            <svg
                              class="mr-2"
                              fill="currentColor"
                              height="1em"
                              stroke="currentColor"
                              stroke-width="0"
                              viewBox="0 0 24 24"
                              width="1em"
                              xmlns="http://www.w3.org/2000/svg"
                            >
                              <path
                                d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
                              />
                            </svg>
                            <div>
                              Close
                            </div>
                          </div>
                        </button>
                      </div>
                    </div>
                  </div>
                </div>
              </div>
            </div>
            <div
              class="pl-1 w-50"
            >
              <div
                class="modalInstallationWrapper"
              >
                <div
                  class="position-relative buttonWrapper"
                >
                  <button
                    class="font-weight-bold text-uppercase position-relative btn btn-block btn btn-outline-secondary btn-sm"
                    data-testid="openModalBtn"
                    type="button"
                  >
                    <div
                      class="d-flex align-items-center justify-content-center"
                    >
                      <svg
                        class="mr-2"
                        fill="none"
                        height="1em"
                        stroke="currentColor"
                        stroke-linecap="round"
                        stroke-linejoin="round"
                        stroke-width="2"
                        viewBox="0 0 24 24"
                        width="1em"
                        xmlns="http://www.w3.org/2000/svg"
                      >
                        <path
                          d="M21 15v4a2 2 0 0 1-2 2H5a2 2 0 0 1-2-2v-4"
                        />
                        <polyline
                          points="7 10 12 15 17 10"
                        />
                        <line
                          x1="12"
                          x2="12"
                          y1="15"
                          y2="3"
                        />
                      </svg>
                      <span>
                        Install
                      </span>
                    </div>
                  </button>
                </div>
                <div
                  class="modal modal"
                  role="dialog"
                >
                  <div
                    class="modal-dialog modal-lg modal-dialog-centered modal-dialog-scrollable"
                  >
                    <div
                      class="modal-content content undefined"
                    >
                      <div
                        class="modal-header d-flex flex-row align-items-center header undefined"
                      >
                        <div
                          class="d-flex align-items-center"
                        >
                          <div
                            class="d-flex align-items-center justify-content-center p-1 overflow-hidden imageWrapper imageWrapper"
                          >
                            <img
                              alt="Pretty name"
                              class="image"
                              src="/image/imageId"
                              srcset="/image/imageId@1x 1x, /image/imageId@2x 2x, /image/imageId@3x 3x, /image/imageId@4x 4x"
                            />
                          </div>
                          <div
                            class="ml-3"
                          >
                            <div
                              class="h5 mb-0"
                            >
                              Pretty name
                            </div>
                          </div>
                        </div>
                        <button
                          class="close"
                          data-testid="closeModalBtn"
                          type="button"
                        >
                          <span
                            aria-hidden="true"
                          >
                            ×
                          </span>
                        </button>
                      </div>
                      <div
                        class="modal-body p-4 h-100 d-flex flex-column"
                      >
                        <div
                          class="alertWrapper"
                          data-testid="alertWrapper"
                        />
                      </div>
                      <div
                        class="modal-footer p-3"
                      >
                        <button
                          class="btn btn-sm btn-secondary text-uppercase"
                          data-testid="closeModalFooterBtn"
                          type="button"
                        >
                          <div
                            class="d-flex flex-row align-items-center"
                          >
                            <svg
                              class="mr-2"
                              fill="currentColor"
                              height="1em"
                              stroke="currentColor"
                              stroke-width="0"
                              viewBox="0 0 24 24"
                              width="1em"
                              xmlns="http://www.w3.org/2000/svg"
                            >
                              <path
                                d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
                              />
                            </svg>
                            <div>
                              Close
                            </div>
                          </div>
                        </button>
                      </div>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="container"
    >
      <div
        class="row"
      >
        <div
          class="mainContent"
        >
          <div
            class="alert alert-primary ml-auto mr-auto my-5 text-center p-4 p-sm-5 border wrapper"
            data-testid="noData"
          >
            <div
              class="h4"
            >
              No README file available for this package
            </div>
          </div>
        </div>
        <div
          class="col col-auto pl-5 pb-4 d-none d-md-block"
        >
          <div
            class="modalInstallationWrapper"
          >
            <div
              class="position-relative buttonWrapper"
            >
              <button
                class="font-weight-bold text-uppercase position-relative btn btn-block btn btn-primary"
                data-testid="openModalBtn"
                type="button"
              >
                <div
                  class="d-flex align-items-center justify-content-center"
                >
                  <svg
                    class="mr-2"
                    fill="none"
                    height="1em"
                    stroke="currentColor"
                    stroke-linecap="round"
                    stroke-linejoin="round"
                    stroke-width="2"
                    viewBox="0 0 24 24"
                    width="1em"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M21 15v4a2 2 0 0 1-2 2H5a2 2 0 0 1-2-2v-4"
                    />
                    <polyline
                      points="7 10 12 15 17 10"
                    />
                    <line
                      x1="12"
                      x2="12"
                      y1="15"
                      y2="3"
                    />
                  </svg>
                  <span>
                    Install
                  </span>
                </div>
              </button>
            </div>
            <div
              class="modal modal"
              role="dialog"
            >
              <div
                class="modal-dialog modal-lg modal-dialog-centered modal-dialog-scrollable"
              >
                <div
                  class="modal-content content undefined"
                >
                  <div
                    class="modal-header d-flex flex-row align-items-center header undefined"
                  >
                    <div
                      class="d-flex align-items-center"
                    >
                      <div
                        class="d-flex align-items-center justify-content-center p-1 overflow-hidden imageWrapper imageWrapper"
                      >
                        <img
                          alt="Pretty name"
                          class="image"
                          src="/image/imageId"
                          srcset="/image/imageId@1x 1x, /image/imageId@2x 2x, /image/imageId@3x 3x, /image/imageId@4x 4x"
                        />
                      </div>
                      <div
                        class="ml-3"
                      >
                        <div
                          class="h5 mb-0"
                        >
                          Pretty name
                        </div>
                      </div>
                    </div>
                    <button
                      class="close"
                      data-testid="closeModalBtn"
                      type="button"
                    >
                      <span
                        aria-hidden="true"
                      >
                        ×
                      </span>
                    </button>
                  </div>
                  <div
                    class="modal-body p-4 h-100 d-flex flex-column"
                  >
                    <div
                      class="alertWrapper"
                      data-testid="alertWrapper"
                    />
                  </div>
                  <div
                    class="modal-footer p-3"
                  >
                    <button
                      class="btn btn-sm btn-secondary text-uppercase"
                      data-testid="closeModalFooterBtn"
                      type="button"
                    >
                      <div
                        class="d-flex flex-row align-items-center"
                      >
                        <svg
                          class="mr-2"
                          fill="currentColor"
                          height="1em"
                          stroke="currentColor"
                          stroke-width="0"
                          viewBox="0 0 24 24"
                          width="1em"
                          xmlns="http://www.w3.org/2000/svg"
                        >
                          <path
                            d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
                          />
                        </svg>
                        <div>
                          Close
                        </div>
                      </div>
                    </button>
                  </div>
                </div>
              </div>
            </div>
          </div>
          <div
            class="card shadow-sm position-relative info info"
          >
            <div
              class="card-body detailsBody"
            >
              <div>
                <div
                  class="mt-2 mb-1 undefined"
                  data-testid="smallTitle"
                >
                  <small
                    class="card-title text-muted text-uppercase"
                  >
                    <span>
                      Application version
                    </span>
                  </small>
                </div>
                <p
                  class="text-truncate"
                  data-testid="appVersion"
                >
                  1.0.0
                </p>
              </div>
              <div>
                <div
                  class="d-flex flex-row align-items-center mt-2 mb-1"
                >
                  <div
                    data-testid="smallTitle"
                  >
                    <small
                      class="card-title text-muted text-uppercase"
                    >
                      Chart Versions
                    </small>
                  </div>
                  <small>
                    <a
                      class="badge badge-pill badge-secondary rssBadge ml-3 badge"
                      href="/api/v1/packages/helm/incubator/pr/feed/rss"
                      rel="alternate noopener noreferrer"
                      role="button"
                      target="_blank"
                      type="application/rss+xml"
                    >
                      <div
                        class="d-flex flex-row align-items-center"
                      >
                        <svg
                          class="mr-1"
                          fill="currentColor"
                          height="1em"
                          stroke="currentColor"
                          stroke-width="0"
                          viewBox="0 0 448 512"
                          width="1em"
                          xmlns="http://www.w3.org/2000/svg"
                        >
                          <path
                            d="M128.081 415.959c0 35.369-28.672 64.041-64.041 64.041S0 451.328 0 415.959s28.672-64.041 64.041-64.041 64.04 28.673 64.04 64.041zm175.66 47.25c-8.354-154.6-132.185-278.587-286.95-286.95C7.656 175.765 0 183.105 0 192.253v48.069c0 8.415 6.49 15.472 14.887 16.018 111.832 7.284 201.473 96.702 208.772 208.772.547 8.397 7.604 14.887 16.018 14.887h48.069c9.149.001 16.489-7.655 15.995-16.79zm144.249.288C439.596 229.677 251.465 40.445 16.503 32.01 7.473 31.686 0 38.981 0 48.016v48.068c0 8.625 6.835 15.645 15.453 15.999 191.179 7.839 344.627 161.316 352.465 352.465.353 8.618 7.373 15.453 15.999 15.453h48.068c9.034-.001 16.329-7.474 16.005-16.504z"
                          />
                        </svg>
                        <div>
                          RSS
                        </div>
                      </div>
                    </a>
                  </small>
                </div>
                <p
                  data-testid="chartVersions"
                >
                  -
                </p>
              </div>
              <div
                class="mt-2 mb-1 undefined"
                data-testid="smallTitle"
              >
                <small
                  class="card-title text-muted text-uppercase"
                >
                  <span>
                    Keywords
                  </span>
                </small>
              </div>
              <div
                class="mb-3"
              >
                <span
                  data-testid="keywords"
                >
                  <button
                    class="btn btn-sm d-inline badge font-weight-normal mr-2 mb-2 mb-sm-0 mw-100 badge"
                    data-testid="keywordBtn"
                  >
                    <div
                      class="text-truncate"
                    >
                      key1
                    </div>
                  </button>
                  <button
                    class="btn btn-sm d-inline badge font-weight-normal mr-2 mb-2 mb-sm-0 mw-100 badge"
                    data-testid="keywordBtn"
                  >
                    <div
                      class="text-truncate"
                    >
                      key2
                    </div>
                  </button>
                </span>
              </div>
            </div>
          </div>
          <div
            class="relatedPackagesWrapper"
          />
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
